File: problem_14.rb
#Which starting number, under a given number, produces the longest Collatz sequence?
module CollatzConjecture

  def self.max_steps_in_range(top_range)
    max_steps = 0
    number = 0
    (1...top_range).each do |candidate|
      if steps(candidate) > max_steps
        max_steps = steps(candidate)
        number = candidate
      end
    end
    puts "The answer is %d with %d steps." % [number, max_steps]
  end

  def self.steps(number)
    raise ArgumentError if number <= 0
    count = 0
    until number == 1
      count += 1
      number = number.even? ? number / 2 : number * 3 + 1
    end
    count
  end

end

if __FILE__ == $PROGRAM_NAME
  CollatzConjecture.max_steps_in_range(1_000_000)
end

File: problem_14.py
def main():
    max_steps = 0
    number = 0
    for candidate in range(1, 1_000_000):
        if steps(candidate) > max_steps:
            max_steps = steps(candidate)
            number = candidate
    print("The answer is", number, "with", max_steps, "steps.")

def steps(number, count = 0):
    if number <= 0:
        raise ValueError("Only positive integers are allowed")
    if number == 1:
        return count
    else:
        if number % 2 == 0:
            count += 1
            return steps(number / 2, count)
        else:
            count += 1
            return steps(number * 3 + 1, count)

if __name__ == "__main__":
    main()
